File: solutions/go/house/1/house.go
package house

import "fmt"

var subjects = []string{
	"house",
	"malt",
	"rat",
	"cat",
	"dog",
	"cow with the crumpled horn",
	"maiden all forlorn",
	"man all tattered and torn",
	"priest all shaven and shorn",
	"rooster that crowed in the morn",
	"farmer sowing his corn",
	"horse and the hound and the horn",
}

var actions = []string{
	"that Jack built.",
	"that lay in",
	"that ate",
	"that killed",
	"that worried",
	"that tossed",
	"that milked",
	"that kissed",
	"that married",
	"that woke",
	"that kept",
	"that belonged to",
}

func buildLine(idx int) string {
	if idx == 0 {
		return fmt.Sprintf("that Jack built.")
	}

	if idx == 1 {
		return fmt.Sprintf("%s the %s %s", actions[idx], subjects[idx-1], buildLine(idx-1))
	}

	return fmt.Sprintf("%s the %s\n%s", actions[idx], subjects[idx-1], buildLine(idx-1))
}

func Verse(v int) string {
	if v == 1 {
		return fmt.Sprintf("This is the house that Jack built.")
	}
	return fmt.Sprintf("This is the %s\n%s", subjects[v-1], buildLine(v-1))
}

func Song() string {
	song := Verse(1)
	for i := 2; i <= len(subjects); i++ {
		song += "\n\n" + Verse(i)
	}
	return song
}

File: solutions/typescript/house/1/house.ts
export function verse(verseNumber: number): string[] {
  if (verseNumber === 1) {
    return ['This is the house that Jack built.']
  }
  if (verseNumber === 2) {
    return [
      'This is the malt',
      'that lay in the house that Jack built.',
    ]
  }
  if (verseNumber === 3) {
    return [
      'This is the rat',
      'that ate the malt',
      'that lay in the house that Jack built.',
    ]
  }
  if (verseNumber === 4) {
    return [
      'This is the cat',
      'that killed the rat',
      'that ate the malt',
      'that lay in the house that Jack built.',
    ]
  }
  if (verseNumber === 5) {
    return [
      'This is the dog',
      'that worried the cat',
      'that killed the rat',
      'that ate the malt',
      'that lay in the house that Jack built.',
    ]
  }
  if (verseNumber === 6) {
    return [
      'This is the cow with the crumpled horn',
      'that tossed the dog',
      'that worried the cat',
      'that killed the rat',
      'that ate the malt',
      'that lay in the house that Jack built.',
    ]
  }
  if (verseNumber === 7) {
    return [
      'This is the maiden all forlorn',
      'that milked the cow with the crumpled horn',
      'that tossed the dog',
      'that worried the cat',
      'that killed the rat',
      'that ate the malt',
      'that lay in the house that Jack built.',
    ]
  }

  if (verseNumber === 8) {
    return  [
      'This is the man all tattered and torn',
      'that kissed the maiden all forlorn',
      'that milked the cow with the crumpled horn',
      'that tossed the dog',
      'that worried the cat',
      'that killed the rat',
      'that ate the malt',
      'that lay in the house that Jack built.',
    ]
  }

  if (verseNumber === 9) {
    return [
        'This is the priest all shaven and shorn',
        'that married the man all tattered and torn',
        'that kissed the maiden all forlorn',
        'that milked the cow with the crumpled horn',
        'that tossed the dog',
        'that worried the cat',
        'that killed the rat',
        'that ate the malt',
        'that lay in the house that Jack built.',
      ]
    }

    if (verseNumber === 10) {
      return [
        'This is the rooster that crowed in the morn',
        'that woke the priest all shaven and shorn',
        'that married the man all tattered and torn',
        'that kissed the maiden all forlorn',
        'that milked the cow with the crumpled horn',
        'that tossed the dog',
        'that worried the cat',
        'that killed the rat',
        'that ate the malt',
        'that lay in the house that Jack built.',
      ]
    }

    if (verseNumber === 11) {
      return  [
        'This is the farmer sowing his corn',
        'that kept the rooster that crowed in the morn',
        'that woke the priest all shaven and shorn',
        'that married the man all tattered and torn',
        'that kissed the maiden all forlorn',
        'that milked the cow with the crumpled horn',
        'that tossed the dog',
        'that worried the cat',
        'that killed the rat',
        'that ate the malt',
        'that lay in the house that Jack built.',
      ]
    }

    if (verseNumber === 12) {
      return [
        'This is the horse and the hound and the horn',
        'that belonged to the farmer sowing his corn',
        'that kept the rooster that crowed in the morn',
        'that woke the priest all shaven and shorn',
        'that married the man all tattered and torn',
        'that kissed the maiden all forlorn',
        'that milked the cow with the crumpled horn',
        'that tossed the dog',
        'that worried the cat',
        'that killed the rat',
        'that ate the malt',
        'that lay in the house that Jack built.',
      ]
    }

  return []
}

export function verses(startVerse: number, endVerse: number): string[] {
  let lyrics: string[] = []

  for (let i = startVerse; i <= endVerse; i++) {
    lyrics = lyrics.concat(verse(i))
    if (i < endVerse) {
      lyrics.push('')
    }
  }

    return lyrics
}
